src/server/clients_info/client_visibility.rs

```diff
@@ -0,0 +1,545 @@
+use bevy::{
+    prelude::*,
+    utils::{EntityHashMap, EntityHashSet},
+};
+
+use super::VisibilityPolicy;
+
+/// Entity visibility settings for a client.
+pub struct ClientVisibility {
+    filter: VisibilityFilter,
+
+    /// Visibility for a specific entity that has been cached for re-referencing.
+    ///
+    /// Used as an optimization by server replication.
+    cached_visibility: Visibility,
+}
+
+impl ClientVisibility {
+    /// Creates a new instance based on the preconfigured policy.
+    pub(super) fn new(policy: VisibilityPolicy) -> Self {
+        match policy {
+            VisibilityPolicy::All => Self::with_filter(VisibilityFilter::All {
+                just_connected: true,
+            }),
+            VisibilityPolicy::Blacklist => Self::with_filter(VisibilityFilter::Blacklist {
+                list: Default::default(),
+                added: Default::default(),
+                removed: Default::default(),
+            }),
+            VisibilityPolicy::Whitelist => Self::with_filter(VisibilityFilter::Whitelist {
+                list: Default::default(),
+                added: Default::default(),
+                removed: Default::default(),
+            }),
+        }
+    }
+
+    /// Creates a new instance with a specific filter.
+    fn with_filter(filter: VisibilityFilter) -> Self {
+        Self {
+            filter,
+            cached_visibility: Default::default(),
+        }
+    }
+
+    /// Resets the filter state to as it was after [`Self::new`].
+    ///
+    /// `cached_visibility` remains untouched.
+    pub(super) fn clear(&mut self) {
+        match &mut self.filter {
+            VisibilityFilter::All { just_connected } => *just_connected = true,
+            VisibilityFilter::Blacklist {
+                list,
+                added,
+                removed,
+            } => {
+                list.clear();
+                added.clear();
+                removed.clear();
+            }
+            VisibilityFilter::Whitelist {
+                list,
+                added,
+                removed,
+            } => {
+                list.clear();
+                added.clear();
+                removed.clear();
+            }
+        }
+    }
+
+    /// Updates list information and its sets based on the filter.
+    ///
+    /// Should be called after each tick.
+    pub(crate) fn update(&mut self) {
```

Comment:
Since we use drain now, maybe it worth to combine update and drain logic to avoid the confusion about cleanup?